[Proposal] Improve the DATA_LIMIT atribute to be able to handle more uses cases

QuentinDuval opened the pull request:
This PR is a draft, pushed for visibility and discussion.

The additional uses cases I propose to support are:
- being able to sub-select part of a dataset in a balanced way (each label is included the same number of time)
- being able to sub-select exclusive parts of the same dataset (for instance to have a validation set that does not intersect with a training set, useful for HP searches)
- make sure that this sub-sampling is deterministic (same seed across all distributed workers)

This would avoid having to create sub-sets of datasets such as ImageNet to test on 1% of each label for instance. It would also allow to benchmark SSL algorithms on low data regime in a more flexible way.

/!\ This PR introduces a breaking change (DATA_LIMIT is not an integer anymore but a structure)

This PR includes:
- unit tests for the sub-sampling strategies
- update of all configuration using the DATA_LIMIT attribute

prigoyal reviewed (approved):
super cool and makes a lot of sense IMO!! thank you @QuentinDuval :)

one small nit: `NB_SAMPLES` -> `NUM_SAMPLES` . We use `num` everywhere in VISSL. consistency enforcement!

there are some bash scripts and possibly docs that might require update with this PR. can you grep once to double check? I love this change :)

QuentinDuval commented:
> super cool and makes a lot of sense IMO!! thank you @QuentinDuval :)
> 
> one small nit: `NB_SAMPLES` -> `NUM_SAMPLES` . We use `num` everywhere in VISSL. consistency enforcement!
> 
> there are some bash scripts and possibly docs that might require update with this PR. can you grep once to double check? I love this change :)

Makes sense :) I will rename the attribute and look inside the docs (I searched the code, but I forgot to search the doc so there is likely some change to be made)

QuentinDuval commented:
> super cool and makes a lot of sense IMO!! thank you @QuentinDuval :)
> 
> one small nit: `NB_SAMPLES` -> `NUM_SAMPLES` . We use `num` everywhere in VISSL. consistency enforcement!
> 
> there are some bash scripts and possibly docs that might require update with this PR. can you grep once to double check? I love this change :)

Renamed to NUM_SAMPLES, and corrected the references, following d394af5 :)

prigoyal commented:
hmm the circleCI gpu error triggered is very interesting. Let's investigate that. We can also try re-running the test in case it was a flaky instance.

QuentinDuval commented:
> hmm the circleCI gpu error triggered is very interesting. Let's investigate that. We can also try re-running the test in case it was a flaky instance.

I will look into this, it might have broken some configurations.

prigoyal commented on vissl/data/data_helper.py:
```diff
@@ -24,6 +24,65 @@ def get_mean_image(crop_size):
+    Context manager to run a specific portion of code with a given seed:
+    resumes the numpy state after the execution of the block
+    """
+    random_state = np.random.get_state()
```
rename to `orig_random_state` :)

QuentinDuval replied:
It will be done :)

prigoyal commented on vissl/data/data_helper.py:
```diff
@@ -24,6 +24,65 @@ def get_mean_image(crop_size):
 
 
+@contextlib.contextmanager
+def with_temporary_numpy_seed(seed: int):
```
nit: rename `sampling_seed` for clarity that it's not the other training seed :)

prigoyal commented on vissl/data/data_helper.py:
```diff
@@ -24,6 +24,65 @@ def get_mean_image(crop_size):
+
+
+def unbalanced_sub_sampling(
+    total_size: int, nb_samples: int, skip_samples: int = 0, seed: int = 0
```
nit: `total_size` -> `total_num_samples`

prigoyal commented on vissl/data/data_helper.py:
```diff
@@ -24,6 +24,65 @@ def get_mean_image(crop_size):
+        groups.setdefault(label, []).append(i)
+
+    unique_labels = sorted(groups.keys())
+    sq, sr = divmod(skip_samples, len(unique_labels))
+    q, r = divmod(nb_samples, len(unique_labels))
```
is it feasible to make the `sq, sr, q, r` more expanded names? :)

QuentinDuval replied:
Indeed... sorry for that. I wanted to do it, but it slipped my mind and forgot to do it. I will fix this !

prigoyal commented on vissl/data/data_helper.py:
```diff
@@ -24,6 +24,65 @@ def get_mean_image(crop_size):
+    Return the selected indices
+    """
+    groups = {}
+    for i, label in enumerate(labels):
```
does this have any assumption on the labels type being list, numpy array etc ?

QuentinDuval replied:
The type hints says that the "labels" should be a numpy array, but this could be relaxed to any kind of Sequence. The other assumption is that the labels inside it should be hashable.

prigoyal replied:
docstring is perfect :) also , since labels can be many things, we can remove the type hint for labels or use the `Union` to compose the typehint :)

prigoyal commented on vissl/data/ssl_dataset.py:
```diff
@@ -241,11 +265,17 @@ def __getitem__(self, idx):
             self._load_labels()
             self._labels_init = True
 
+        if self.data_limit >= 0:
+            self._init_subset()
```
can we add the check `if not self._subset_initialized`: right here? so that it clear that we only call the init_subset if needed ? :)

prigoyal commented on vissl/data/ssl_dataset.py:
```diff
@@ -241,11 +265,17 @@ def __getitem__(self, idx):
             self._load_labels()
             self._labels_init = True
 
+        if self.data_limit >= 0:
+            self._init_subset()
+            subset_idx = self.sub_set[idx]
+        else:
+            subset_idx = idx
```
we could write this whole as
```
item_index = idx
if self.data_limit >= 0:
    if not self._subset_initialized:
         self._init_subset()
    item_index = self.subset[idx]
```

what do you think?

prigoyal commented on vissl/data/disk_dataset.py:
```diff
@@ -87,13 +87,6 @@ def _load_data(self, path):
             # Avoid creating it over and over again.
             self.is_initialized = True
 
-        if self.cfg["DATA"][self.split]["DATA_LIMIT"] > 0:
```
I am a little bit worried about moving this out to the ssl_dataset.py. The reason is because we have some internal data sources that don't apply the data limit in similar way and could break training for them. Since our intent is to support the data limit like this for the synthetic and disk_dataset, can we put the `init_subset()` in data helper and call the function from respective disk_dataset.py and `synthetic_dataset.py` ? any other thoughts/suggestions. Sorry if I am missing some piece :)

QuentinDuval replied:
Do you have some pointers to these internal data sources ?

I mostly moved the code to `ssl_dataset.py` for two reasons:
- I saw that everything was scheduled in `ssl_dataset` (but as you point out, I clearly missing something) and found it would be a good way to factorise the code and the feature (we cannot random sample twice)
- The balanced random sampling (same amount of sample by classes) requires to have access to the labels, and these labels are only available in ssl_dataset

prigoyal replied:
yes the suggestion was to remove it move it to synthetic_dataset and disk_dataset. Let me share some examples . 
1) Essentially https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/fbcode/deeplearning/projects/ssl_framework/vissl/fb/data/fairstore_dataset.py?commit=beaa5fafc669889804877ed3efd62c9df87e1507&lines=103-104 doesn't have the concept of DATA_LIMIT 
2) an internal change in https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/fbcode/deeplearning/projects/ssl_framework/vissl/fb/data/everstore_dataset.py?commit=beaa5fafc669889804877ed3efd62c9df87e1507&lines=162-165 will be needed (a diff from fbcode) to adapt to the new call :)

prigoyal reviewed (commented):
Thank you @QuentinDuval , looks really good and a very  nice proposal too :)

I left some inline thoughts . lmk what you think :) we can aim to land this soon within a day.

QuentinDuval commented on vissl/data/synthetic_dataset.py:
```diff
@@ -19,15 +19,16 @@ class SyntheticImageDataset(Dataset):
     """
 
-    def __init__(self, cfg, path, split, dataset_name, data_source="synthetic"):
+    DEFAULT_SIZE = 50_000
```
I thought 50_000 would be more readable than 50000.

I checked and the associated PEP that introduces this feature in Python (https://www.python.org/dev/peps/pep-0515/) has been integrated in Python 3.6 (so it should work with all VISSL supported Python versions).

prigoyal replied:
oh nice! this is amazing and I learn something! thank you Quentin :)

QuentinDuval commented on vissl/data/ssl_dataset.py:
```diff
@@ -60,20 +62,24 @@ class GenericSSLDataset(Dataset):
         self.dataset_names = self.cfg["DATA"][split].DATASET_NAMES
         self.label_type = self.cfg["DATA"][split].LABEL_TYPE
+        self.data_limit = self.cfg["DATA"][split].DATA_LIMIT
+        self.data_limit_sampling = self.cfg["DATA"][split].DATA_LIMIT_SAMPLING
```
Shouldn't the "default.yaml" make sure that there will always be such a node?

In any case, I can indeed use ".get()", in which case the best is probably to return a default data-structure, with the options such as "not balanced" and "seed = 0" and "skip = 0". What do you think?

prigoyal replied:
yep that sounds good. And I also agree that `default.yaml` should take care of it. However, I discovered over the last few days that some projects are using VISSL and somehow their code breaks if we don't use .get(). It's unclear that if they use new vissl code they why their defaults.yaml is not getting updated. Something to dig into later separately for me.

QuentinDuval commented on vissl/data/ssl_dataset.py:
```diff
@@ -60,20 +62,24 @@ class GenericSSLDataset(Dataset):
         self.label_objs = []
         self.data_paths = []
         self.label_paths = []
+        self.sub_set = None
```
Oh, it will affect both the images and the labels: it effectively reduces the size of the dataset by mapping the range [0, K) to [0, N), where K < N is the DATA_LIMIT value.

QuentinDuval commented on vissl/data/ssl_dataset.py:
```diff
@@ -221,7 +227,25 @@ def _load_labels(self):
+        if not self._subset_initialized:
+            if not self.data_limit_sampling.BALANCED:
+                self.sub_set = unbalanced_sub_sampling(
+                    total_size=len(self.data_objs[0]),
```
For the moment, there are a few that seems to be made with the assumption that we have 1 data source, or that all data sources have the same length. I think this code does not make further assumption (*).

I am actually not quite sure where to push this in the future. The way I see it, we need to rework a bit the way we combine datasets. In particular, combining several datasets (of different length) means reworking the `ssl_dataset.py` module such that:
- combining means concatenating (which is different from the current behaviour, which, if we have 2 dataset, is such that we draw two samples, one for each dataset, at each call to __getitem__)
- labels of different kind can be concatenated (meaning that 2 datasets with different labels means twice as much labels, and optionally deal with the overlap?)

(*) In the current state of the code, the `_init_subset` function does not actually requires to have just one dataset: we can have several, as long as they have the same length, but the sampling will be performed on the first one (which means that the "balanced" constraint might not be respected for the second source, but overall the code will run fine).

What we can do is add an assert for the balanced in case we have either "no label" or "more than 1 data source".

QuentinDuval commented on vissl/config/defaults.yaml:
```diff
@@ -250,8 +250,25 @@ config:
+      DATA_LIMIT_SAMPLING:
+        SEED: 0
+        BALANCED: False
+        SKIP: 0
```
100% agree, the current name is misleading, I will rename this :)

QuentinDuval commented on vissl/config/defaults.yaml:
```diff
@@ -250,8 +250,25 @@ config:
+      #
+      DATA_LIMIT_SAMPLING:
+        SEED: 0
+        BALANCED: False
```
I think `IS_BALANCED` should suffices as the `SAMPLING` part is already implied by the father node. So if a user wants to use the override syntax, the command line will already contain it:

`config.DATA.TRAIN.DATA_LIMIT_SAMPLING.IS_BALANCED=True`

What do you think?

prigoyal replied:
sounds good :)

prigoyal commented:
looks great and good to go! just 2 open comments: 1) use of `.get()` 2) renaming the new config variables (IS_BALANCED) :)

QuentinDuval commented:
> looks great and good to go! just 2 open comments: 1) use of `.get()` 2) renaming the new config variables (IS_BALANCED) :)

Indeed, I really need to fix these things and I will see to it Monday :-) Thank you @prigoyal for your review !

QuentinDuval commented:
> > looks great and good to go! just 2 open comments: 1) use of `.get()` 2) renaming the new config variables (IS_BALANCED) :)
> 
> Indeed, I really need to fix these things and I will see to it Monday :-) Thank you @prigoyal for your review !

Both issues should be fixed at 6c4788e :)

prigoyal commented on vissl/data/ssl_dataset.py:
```diff
@@ -221,7 +234,53 @@ def _load_labels(self):
+        or that all data sources have the same length (same as __getitem__).
+        """
+
+        valid_datasets = {
```
I think we should move this check down to line 305 . We should just not call the init_subset function if the conditions are not met.

QuentinDuval replied:
It should be fixed at 5586fac !

prigoyal reviewed (commented):
one small comment inline and good to go otherwise! :) :)

prigoyal commented:
the fvcore issue has struck here too. Can you rebase on master and push again? A recent landed PR will help with the circleCI tests.
